File: LongestSubString.js
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> set = new HashSet<>();
        int maxLength=0, slow=0, fast=0;
        while(fast<s.length() && slow< s.length()){
            if(!set.contains(s.charAt(fast))){
                set.add(s.charAt(fast));
                fast++;
                maxLength= Math.max(maxLength,fast-slow);
            }else{
                set.remove(s.charAt(slow));
                slow++;
            }
        }
            return maxLength;
    }
}

File: longest-sub-string.java
//Time complexity: O(N)
//Space complexity: O(min(M, N))
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<>();
        int max = 0, slow = 0;
        for(int i = 0 ; i < s.length(); i++) {
            if(map.containsKey(s.charAt(i))) {
                slow = Math.max(slow, map.get(s.charAt(i)));
            }
            max = Math.max(max,  i - slow + 1);
            map.put(s.charAt(i), i+1);
        }
        return max;
    }
}

